crates/iota-types/src/move_package.rs

```diff
@@ -98,6 +104,56 @@ pub struct FnInfoKey {
 /// A map from function info keys to function info
 pub type FnInfoMap = BTreeMap<FnInfoKey, FnInfo>;
 
+/// Create a function info map from function infos
+/// functions infos contain information about functions and their attributes in
+/// a module
+pub fn create_fn_info_map(
+    mod_addr: AccountAddress,
+    mod_is_test: bool,
+    fn_infos: &UniqueMap<FunctionName, FunctionInfo>,
+) -> FnInfoMap {
+    let mut fn_info_map: FnInfoMap = BTreeMap::new();
+
+    for (_, name, info) in fn_infos.iter() {
+        let fn_name = name.as_str().to_string();
+        let is_test = mod_is_test || info.attributes.is_test_or_test_only();
+        let authenticator_version = info.attributes.get_authenticator();
+
+        fn_info_map.insert(
+            FnInfoKey { fn_name, mod_addr },
+            FnInfo {
+                is_test,
+                authenticator_version,
+            },
+        );
+    }
+    fn_info_map
+}
+
+/// Fill runtime metadata for a single compiled module.
+pub fn fill_module_with_metadata(module: &mut CompiledModule, fn_info_map: &FnInfoMap) {
+    let mut runtime_metadata = RuntimeModuleMetadata::default();
+
+    for fn_def in &module.function_defs {
+        let fn_handle = module.function_handle_at(fn_def.function);
+        let fn_name = module.identifier_at(fn_handle.name);
+
+        if let Some(version) = get_authenticator_version_from_fun(fn_name, module, fn_info_map) {
+            runtime_metadata.add_function_attribute(
+                fn_name.to_string(),
+                IotaAttribute::authenticator_attribute(version),
+            );
+        }
+    }
+
+    if !runtime_metadata.is_empty() {
+        module.metadata.push(Metadata {
```

Comment:
`module.metadata` is a vector. We wouldn't really realize here that `IOTA_METADATA_KEY` might already exist in that `metadata`, right?. Would it make sense to add a function that loops over all entries in `metadata` first, before appending to the end? Or are we sure that can never ever happen? If this function here for example is called twice, we have two metadata with the same key, right?